### PR TITLE
Stop stdout & stderr after timeout

### LINKED ISSUE #23: Potential write to WebSocket after close
During my tests, I noticed some warning messages (and probably errors) after an execution was stopped due to the permitted execution time (10 seconds in the example below). When the runner is not destroyed immediately by CodeOcean (which I disabled for testing), the system still sends messages to a closed WebSocket connection (which one?).

```
time="2021-10-31T00:53:18.114276Z" level=debug code=204 duration=226.75ms method=PATCH path=/api/v1/runners/3-e3d2df46-39d3-11ec-b90f-acde48001122/files user_agent="Faraday v1.8.0"
time="2021-10-31T00:53:18.116492Z" level=debug code=200 duration="43µs" method=POST path=/api/v1/runners/3-e3d2df46-39d3-11ec-b90f-acde48001122/execute user_agent="Faraday v1.8.0"
time="2021-10-31T00:53:18.118108Z" level=info msg="Running execution" executionID=160a9884-f0ae-4600-9830-a66536d29c9b package=api runnerId=3-e3d2df46-39d3-11ec-b90f-acde48001122
time="2021-10-31T00:53:28.118509Z" level=warning msg="Could not send SIGQUIT because nothing was written" package=runner runner=3-e3d2df46-39d3-11ec-b90f-acde48001122
time="2021-10-31T00:53:28.118511Z" level=info msg="Execution returned" package=api
time="2021-10-31T00:53:28.118682Z" level=debug code=200 duration=10.000633s method=GET path=/api/v1/runners/3-e3d2df46-39d3-11ec-b90f-acde48001122/websocket user_agent=
time="2021-10-31T00:53:28.121237Z" level=warning msg="Error writing the exit message" error="websocket: close sent" package=api
time="2021-10-31T00:53:28.121281Z" level=warning msg="Error during websocket close" error="websocket: close sent" package=api
time="2021-10-31T00:53:28.121960Z" level=warning msg="Error writing the exit message" error="websocket: close sent" package=api
time="2021-10-31T00:53:28.121992Z" level=warning msg="Error during websocket close" error="websocket: close sent" package=api
time="2021-10-31T00:53:28.122048Z" level=warning msg="Error writing the exit message" error="websocket: close sent" package=api
time="2021-10-31T00:53:28.122064Z" level=warning msg="Error during websocket close" error="websocket: close sent" package=api
time="2021-10-31T00:53:28.122497Z" level=warning msg="Error writing the exit message" error="websocket: close sent" package=api
time="2021-10-31T00:53:28.122541Z" level=warning msg="Error during websocket close" error="websocket: close sent" package=api
time="2021-10-31T00:53:28.122975Z" level=warning msg="Error writing the exit message" error="websocket: close sent" package=api
time="2021-10-31T00:53:28.123006Z" level=warning msg="Error during websocket close" error="websocket: close sent" package=api
time="2021-10-31T00:53:28.123060Z" level=warning msg="Error writing the exit message" error="websocket: close sent" package=api
time="2021-10-31T00:53:28.123077Z" level=warning msg="Error during websocket close" error="websocket: close sent" package=api
time="2021-10-31T00:53:28.123587Z" level=warning msg="Error writing the exit message" error="websocket: close sent" package=api
time="2021-10-31T00:53:28.123620Z" level=warning msg="Error during websocket close" error="websocket: close sent" package=api
time="2021-10-31T00:53:28.141887Z" level=warning msg="Error writing the exit message" error="websocket: close sent" package=api
time="2021-10-31T00:53:28.141946Z" level=warning msg="Error during websocket close" error="websocket: close sent" package=api
time="2021-10-31T00:53:28.203710Z" level=debug msg="Execution terminated after SIGQUIT" package=runner runner=3-e3d2df46-39d3-11ec-b90f-acde48001122
```

### PR DESCRIPTION
**Pull request by mpass99**
Closes #23

### CODE REVIEW

File: internal/runner/runner.go [hunk 1]
```diff
@@ -224,10 +224,11 @@ func (r *NomadJob) handleExitOrContextDone(ctx context.Context, cancelExecute co
 	// (tty has to be true) and converted to a SIGQUIT signal sent to the foreground process attached to the tty.
 	// By default, SIGQUIT causes the process to terminate and produces a core dump. Processes can catch this signal
 	// and ignore it, which is why we destroy the runner if the process does not terminate after a grace period.
-	n, err := stdin.Write([]byte{SIGQUIT})
-	if n != 1 {
-		log.WithField("runner", r.id).Warn("Could not send SIGQUIT because nothing was written")
-	}
+	_, err := stdin.Write([]byte{SIGQUIT})
+	// if n != 1 {
+	// This is always the case but the SIGQUIT gets received.
```
Comment (MrSerth) [inline internal/runner/runner.go:229]: Hm, this is an unexpected change for me.

Do you mean that the SIGQUIT is definitely sent even though the return value does not match our expectation of one byte? Why is that the case?

Comment (mpass99) [reply internal/runner/runner.go:229]: Thank you for raising this still unclear point.
I agree that this is unusual, when normal behaviour would be more like [this](https://go.dev/play/p/GKSZCdKsOeX).

First of all, I only noticed that the now commented-out warning always appears during executions. Then I confirmed my suspicions with a test:
We add this test snippet in `ExecuteInteractively`:
``` Go
	go r.executeCommand(ctxExecute, command, stdin, stdout, stderr, exitInternal)
	go r.handleExitOrContextDone(ctx, cancelExecute, exitInternal, exit, stdin)

	go func() {
		<- time.After(10 * time.Second)
		n, err := stdin.Write([]byte{SIGQUIT})
		log.WithField("n", n).WithError(err).Info("We just send SIGQUIT")
	}()

	return exit, cancel, nil
```
We then execute sth like: `while true; do sleep 1; echo HI; done`
This results in the log statement:
```
time="2021-11-25T18:41:07.718215Z" level=info msg="We just send SIGQUIT" error="<nil>" n=0 package=runner
```
and the execution to be exited after the 10 seconds... But at this point I do not know more about the background.

Comment (MrSerth) [reply internal/runner/runner.go:229]: Indeed. I was able to confirm it and got the following output in CodeOcean:

```
while true; do sleep 1; echo HI; done

HI
HI
HI
HI
HI
HI
HI
HI
HI
^\
```

Obviously, the character is sent correctly and, as you mentioned, the stopping also works correctly. Regardless of the byte array, the length is always 0. 

Let's add (a link to) this discussion to the comment and accept it for now (if you don't have another idea).

```suggestion
	// The SIGQUIT is sent and correctly processed by the allocation.  However, for an unknown 
	// reason, the number of bytes written is always zero even though the error is nil. 
	// Hence, we disabled this sanity check here. See the MR for more details: 
	// https://github.com/openHPI/poseidon/pull/45#discussion_r757029024
```

**Review by MrSerth — APPROVED** [review]
Thanks, it works like a charm 👌

```
time="2021-11-25T20:25:33.399892Z" level=debug code=200 duration="49µs" method=POST path=/api/v1/runners/31-a5e83658-4e1a-11ec-9d90-acde48001122/execute user_agent="Faraday v1.8.0"
time="2021-11-25T20:25:33.401503Z" level=info msg="Running execution" executionID=86d2674b-d2b1-407b-95a7-0497d69a6f80 package=api runnerId=31-a5e83658-4e1a-11ec-9d90-acde48001122
time="2021-11-25T20:25:48.402015Z" level=info msg="Execution returned" package=api
time="2021-11-25T20:25:48.402262Z" level=debug code=200 duration=15.000809s method=GET path=/api/v1/runners/31-a5e83658-4e1a-11ec-9d90-acde48001122/websocket user_agent=
time="2021-11-25T20:25:48.488409Z" level=debug msg="Execution terminated after SIGQUIT" package=runner runner=31-a5e83658-4e1a-11ec-9d90-acde48001122
```

**Pull request merged**